File: ios/MenCare/Features/Admin/Manage/OperatorEditScreen.swift
import SwiftUI
import Observation

private let defaultDays: Set<DayOfWeek> = [.monday, .tuesday, .wednesday, .thursday, .friday, .saturday]

/// New hire form. Saving creates the operator profile *and* the staff account
/// that lets them sign in — both live behind CatalogRepository.createOperator.
@MainActor
@Observable
final class OperatorEditViewModel {

    private let catalog: CatalogRepository

    var name = ""
    var title = ""
    var email = ""
    var phone = ""
    var selectedServiceIds: Set<String> = []
    var workingDays: Set<DayOfWeek> = defaultDays
    var from = LocalTime(9, 0)
    var to = LocalTime(19, 0)
    /// Chiave del campo → messaggio già localizzato: le regole sono quelle di
    /// `Core/Common/Validation.swift`, non più una seconda copia della regex.
    var fieldErrors: [String: String] = [:]
    var saveError: String?
    var saving = false
    var saved = false

    @ObservationIgnored private var submitted = false

    let state = LoadState()
    private(set) var services: [Service] = []

    init(catalog: CatalogRepository) {
        self.catalog = catalog
    }

    func load() async {
        await state.run { services = try await catalog.services() }
    }

    var canSave: Bool {
        validate().isEmpty && !selectedServiceIds.isEmpty && !workingDays.isEmpty && from < to
    }

    func fieldChanged() {
        saveError = nil
        if submitted { fieldErrors = validate() }
    }

    private func validate() -> [String: String] {
        var errors: [String: String] = [:]
        errors["name"] = validateName(name).message
        errors["email"] = validateEmail(email).message
        // Il telefono è obbligatorio come su POST /catalog/operators: prima non
        // veniva controllato mai e finiva grezzo nel database.
        errors["phone"] = validatePhone(phone).message
        return errors.compactMapValues { $0 }
    }

    func toggleService(_ serviceId: String) {
        if selectedServiceIds.contains(serviceId) {
            selectedServiceIds.remove(serviceId)
        } else {
            selectedServiceIds.insert(serviceId)
        }
    }

    func toggleDay(_ day: DayOfWeek) {
        if workingDays.contains(day) {
            workingDays.remove(day)
        } else {
            workingDays.insert(day)
        }
    }

    func setFrom(_ time: LocalTime) {
        from = time
        if time >= to { to = time.plusMinutes(60) }
    }

    func setTo(_ time: LocalTime) {
        to = time <= from ? from.plusMinutes(60) : time
    }

    func save() {
        guard !saving else { return }
        submitted = true
        let errors = validate()
        if !errors.isEmpty || selectedServiceIds.isEmpty || workingDays.isEmpty || from >= to {
            fieldErrors = errors
            saveError = L("validation_form_invalid")
            return
        }
        fieldErrors = [:]
        saveError = nil
        saving = true
        let phoneToSave = normalizePhone(phone) ?? phone
        Task {
            let shift = [TimeRange(from, to)]
            let result = await catalog.createOperator(
                NewOperator(
                    name: name.trimmingCharacters(in: .whitespaces),
                    title: title.trimmingCharacters(in: .whitespaces),
                    email: normalizeEmail(email),
                    phone: phoneToSave,
                    serviceIds: selectedServiceIds,
                    weeklyHours: Dictionary(uniqueKeysWithValues: workingDays.map { ($0, shift) })
                )
            )
            saving = false
            switch result {
            case .success:
                saved = true
            case .failure(let error):
                if error == .emailAlreadyRegistered {
                    fieldErrors = ["email": L("ops_error_email_taken")]
                } else {
                    saveError = L("validation_save_failed")
                }
            }
        }
    }
}

struct OperatorEditScreen: View {
    @Bindable var viewModel: OperatorEditViewModel
    let onBack: () -> Void

    /// true = opening time, false = closing time.
    @State private var editingFrom: Bool?

    var body: some View {
        VStack(spacing: 0) {
            DarkHeader(contentPadding: EdgeInsets(top: 0, leading: 0, bottom: 8, trailing: 0)) {
                BrandTopBar(
                    title: L("ops_new"),
                    onBack: onBack,
                    backLabel: L("manage_tab_operators")
                ) {
                    BarAction(text: L("svc_save_short"), action: viewModel.save, color: .oliveLight)
                }
            }

            ScrollView {
                VStack(alignment: .leading, spacing: 12) {
                    NameField(
                        label: L("ops_field_name"), text: $viewModel.name,
                        error: viewModel.fieldErrors["name"], outlined: true
                    )
                    .onChange(of: viewModel.name) { viewModel.fieldChanged() }
                    FilledTextField(
                        label: L("ops_field_role"), text: $viewModel.title,
                        placeholder: L("ops_field_role_hint"), autocapitalization: .sentences
                    )
                    EmailField(
                        label: L("ops_field_email"), text: $viewModel.email,
                        error: viewModel.fieldErrors["email"]
                    )
                    .onChange(of: viewModel.email) { viewModel.fieldChanged() }
                    PhoneField(
                        label: L("ops_field_phone"), text: $viewModel.phone,
                        error: viewModel.fieldErrors["phone"]
                    )
                    .onChange(of: viewModel.phone) { viewModel.fieldChanged() }

                    BrandSectionLabel(text: L("ops_working_days"))
                    HStack(spacing: 6) {
                        ForEach(DayOfWeek.allCases, id: \.self) { day in
                            BrandChip(
                                text: formatDayGroup([day]).prefix(3).capitalizedFirst,
                                selected: viewModel.workingDays.contains(day),
                                action: { viewModel.toggleDay(day) },
                                fill: true
                            )
                        }
                    }
                    HStack(spacing: 9) {
                        PickerTile(label: L("business_from"), value: formatTime(viewModel.from)) {
                            editingFrom = true
                        }
                        PickerTile(label: L("business_to"), value: formatTime(viewModel.to)) {
                            editingFrom = false
                        }
                    }

                    BrandSectionLabel(text: L("svc_operators_services"))
                    servicesGrid
                    if viewModel.selectedServiceIds.isEmpty {
                        Text(L("ops_error_services"))
                            .font(Typo.bodySmall)
                            .foregroundStyle(Color.errorRed)
                    }

                    FormErrorBanner(message: viewModel.saveError)

                    Text(L("ops_account_note"))
                        .font(Typo.jost(12))
                        // Su fondo nero il grigio chiaro, non quello da chiaro.
                        .foregroundStyle(Color.onDarkMuted)
                        .padding(14)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .background(RoundedRectangle(cornerRadius: 16).fill(Color.ink))
                }
                .padding(.horizontal, 20)
                .padding(.top, 18)
                .padding(.bottom, 16)
                .readableWidth()
            }

            BottomActionBar {
                AccentButton(
                    text: L("ops_create"),
                    action: viewModel.save,
                    enabled: viewModel.canSave,
                    loading: viewModel.saving,
                    height: 56,
                    corner: 16
                )
                .readableWidth()
            }
        }
        .background(Color.bone)
        .task { await viewModel.load() }
        .onChange(of: viewModel.saved) {
            if viewModel.saved { onBack() }
        }
        .sheet(isPresented: Binding(
            get: { editingFrom != nil },
            set: { if !$0 { editingFrom = nil } }
        )) {
            let isStart = editingFrom ?? true
            BrandTimePickerDialog(
                initial: isStart ? viewModel.from : viewModel.to,
                onDismiss: { editingFrom = nil },
                onConfirm: { time in
                    if isStart { viewModel.setFrom(time) } else { viewModel.setTo(time) }
                    editingFrom = nil
                }
            )
        }
    }

    private var servicesGrid: some View {
        let columns = Array(repeating: GridItem(.flexible(), spacing: 8), count: 2)
        return LazyVGrid(columns: columns, spacing: 8) {
            ForEach(viewModel.services) { service in
                BrandChip(
                    text: service.name,
                    selected: viewModel.selectedServiceIds.contains(service.id),
                    action: { viewModel.toggleService(service.id) },
                    fill: true,
                    maxLines: 2
                )
            }
        }
    }
}

private extension Substring {
    var capitalizedFirst: String {
        prefix(1).uppercased() + dropFirst()
    }
}

File: android/feature/admin/src/main/kotlin/com/devora/mencare/feature/admin/manage/OperatorEditScreen.kt
package com.devora.mencare.feature.admin.manage

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.devora.mencare.core.common.formatTime
import com.devora.mencare.core.designsystem.component.AccentButton
import com.devora.mencare.core.designsystem.component.BarAction
import com.devora.mencare.core.designsystem.component.BottomActionBar
import com.devora.mencare.core.designsystem.component.BrandChip
import com.devora.mencare.core.designsystem.component.BrandSectionLabel
import com.devora.mencare.core.designsystem.component.BrandTimePickerDialog
import com.devora.mencare.core.designsystem.component.BrandTopBar
import com.devora.mencare.core.designsystem.component.DarkHeader
import com.devora.mencare.core.designsystem.component.EmailField
import com.devora.mencare.core.designsystem.component.FilledTextField
import com.devora.mencare.core.designsystem.component.NameField
import com.devora.mencare.core.designsystem.component.PhoneField
import com.devora.mencare.core.designsystem.component.PickerTile
import com.devora.mencare.core.designsystem.component.readableWidth
import com.devora.mencare.core.designsystem.theme.Bone
import com.devora.mencare.core.designsystem.theme.ErrorRed
import com.devora.mencare.core.designsystem.theme.Ink
import com.devora.mencare.core.designsystem.theme.OliveLight
import com.devora.mencare.core.designsystem.theme.OnDarkMuted
import com.devora.mencare.feature.admin.R
import java.time.DayOfWeek
import java.time.format.TextStyle
import java.util.Locale

@Composable
fun OperatorEditScreen(
    onBack: () -> Unit,
    viewModel: OperatorEditViewModel = hiltViewModel(),
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    // true = opening time, false = closing time
    var editingFrom by rememberSaveable { mutableStateOf<Boolean?>(null) }

    LaunchedEffect(state.saved) {
        if (state.saved) onBack()
    }

    Column(modifier = Modifier.fillMaxSize().background(Bone)) {
        DarkHeader(
            contentPadding = PaddingValues(start = 0.dp, end = 0.dp, top = 0.dp, bottom = 8.dp),
        ) {
            BrandTopBar(
                title = stringResource(R.string.ops_new),
                onBack = onBack,
                backLabel = stringResource(R.string.manage_tab_operators),
                trailing = {
                    BarAction(stringResource(R.string.svc_save_short), viewModel::save, color = OliveLight)
                },
            )
        }

        Column(
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .readableWidth()
                .imePadding()
                .padding(horizontal = 20.dp)
                .padding(top = 18.dp, bottom = 16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            NameField(
                value = state.name,
                onValueChange = { viewModel.onFieldChange(OperatorField.NAME, it) },
                label = stringResource(R.string.ops_field_name),
                error = state.errors[OperatorField.NAME],
            )
            FilledTextField(
                value = state.title,
                onValueChange = viewModel::onTitleChange,
                label = stringResource(R.string.ops_field_role),
                placeholder = stringResource(R.string.ops_field_role_hint),
            )
            EmailField(
                value = state.email,
                onValueChange = { viewModel.onFieldChange(OperatorField.EMAIL, it) },
                label = stringResource(R.string.ops_field_email),
                error = state.errors[OperatorField.EMAIL],
                errorText = if (state.emailTaken) stringResource(R.string.ops_error_email_taken) else null,
            )
            PhoneField(
                value = state.phone,
                onValueChange = { viewModel.onFieldChange(OperatorField.PHONE, it) },
                label = stringResource(R.string.ops_field_phone),
                error = state.errors[OperatorField.PHONE],
            )

            BrandSectionLabel(stringResource(R.string.ops_working_days))
            Row(horizontalArrangement = Arrangement.spacedBy(6.dp), modifier = Modifier.fillMaxWidth()) {
                DayOfWeek.entries.forEach { day ->
                    BrandChip(
                        text = day.getDisplayName(TextStyle.SHORT, Locale.ITALIAN)
                            .trimEnd('.')
                            .replaceFirstChar { it.uppercase() },
                        selected = day in state.workingDays,
                        onClick = { viewModel.toggleDay(day) },
                        modifier = Modifier.weight(1f),
                        fill = true,
                    )
                }
            }
            Row(horizontalArrangement = Arrangement.spacedBy(9.dp), modifier = Modifier.fillMaxWidth()) {
                PickerTile(
                    label = stringResource(R.string.business_from),
                    value = formatTime(state.from),
                    onClick = { editingFrom = true },
                    modifier = Modifier.weight(1f),
                )
                PickerTile(
                    label = stringResource(R.string.business_to),
                    value = formatTime(state.to),
                    onClick = { editingFrom = false },
                    modifier = Modifier.weight(1f),
                )
            }

            BrandSectionLabel(stringResource(R.string.svc_operators_services))
            Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                state.services.chunked(2).forEach { row ->
                    Row(horizontalArrangement = Arrangement.spacedBy(8.dp), modifier = Modifier.fillMaxWidth()) {
                        row.forEach { service ->
                            BrandChip(
                                text = service.name,
                                selected = service.id in state.selectedServiceIds,
                                onClick = { viewModel.toggleService(service.id) },
                                modifier = Modifier.weight(1f),
                                fill = true,
                                maxLines = 2,
                            )
                        }
                        if (row.size == 1) Spacer(Modifier.weight(1f))
                    }
                }
            }
            if (state.selectedServiceIds.isEmpty()) {
                Text(
                    stringResource(R.string.ops_error_services),
                    style = MaterialTheme.typography.bodySmall,
                    color = ErrorRed,
                )
            }
            if (state.saveFailed) {
                Text(
                    stringResource(R.string.ops_error_save),
                    style = MaterialTheme.typography.bodySmall,
                    color = ErrorRed,
                )
            }

            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(16.dp))
                    .background(Ink)
                    .padding(14.dp),
            ) {
                Text(
                    stringResource(R.string.ops_account_note),
                    style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp, lineHeight = 18.sp),
                    // Su fondo nero il grigio chiaro, non quello da fondo chiaro.
                    color = OnDarkMuted,
                )
            }
        }

        BottomActionBar(modifier = Modifier.navigationBarsPadding()) {
            AccentButton(
                modifier = Modifier.readableWidth(),
                text = stringResource(R.string.ops_create),
                onClick = viewModel::save,
                enabled = state.canSave,
                loading = state.saving,
                height = 56.dp,
                shape = RoundedCornerShape(16.dp),
            )
        }
    }

    val editing = editingFrom
    if (editing != null) {
        BrandTimePickerDialog(
            initial = if (editing) state.from else state.to,
            onDismiss = { editingFrom = null },
            onConfirm = { time ->
                if (editing) viewModel.setFrom(time) else viewModel.setTo(time)
                editingFrom = null
            },
        )
    }
}
